src/IO/Importers/VolumeDataReaderActions.hpp

```diff
@@ -83,115 +88,130 @@ namespace ThreadedActions {
  * tensor data is stored in datasets named `db::tag_name<Tag>() + suffix`, where
  * the `suffix` is empty for scalars or `"_"` followed by the
  * `Tensor::component_name` for each independent tensor component.
- * - The `CallbackAction` is invoked on each registered element of the
- * `CallbackComponent` with a
- * `tuples::tagged_tuple_from_typelist<FieldTagsList>` containing the
- * tensor data for that element. Use `Actions::SetData` to write the data
- * directly into their respective tags in the DataBox. The `CallbackComponent`
- * must the the same that was encoded into the `observers::ArrayComponentId`
- * used to register the elements.
+ * - `Parallel::receive_data` is invoked on each registered element of the
+ * `ReceiveComponent` to populate `importers::Tags::VolumeData` in the element's
+ * inbox with a `tuples::tagged_tuple_from_typelist<FieldTagsList>` containing
+ * the tensor data for that element. The `ReceiveComponent` must the the same
+ * that was encoded into the `observers::ArrayComponentId` used to register the
+ * elements.
+ *
+ * \see \ref dev_guide_importing
  */
 template <typename ImporterOptionsGroup, typename FieldTagsList,
-          typename CallbackAction, typename CallbackComponent>
-struct ReadVolumeData {
-  using const_global_cache_tags =
-      tmpl::list<Tags::FileName<ImporterOptionsGroup>,
-                 Tags::Subgroup<ImporterOptionsGroup>,
-                 Tags::ObservationValue<ImporterOptionsGroup>>;
+          typename ReceiveComponent>
+struct ReadAllVolumeDataAndDistribute {
+  template <
+      typename ParallelComponent, typename DataBox, typename Metavariables,
+      typename ArrayIndex,
+      Requires<db::tag_is_retrievable_v<Tags::RegisteredElements, DataBox> and
+               db::tag_is_retrievable_v<Tags::HasReadVolumeData, DataBox>> =
+          nullptr>
+  static void apply(DataBox& box, Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/) noexcept {
+    // Only read and distribute the volume data once
+    // This action will be invoked by `importers::Actions::ReadVolumeData` from
+    // every element on the node, but only the first invocation reads the file
+    // and distributes the data to all elements. Subsequent invocations do
+    // nothing. We use the `ImporterOptionsGroup` that specifies the data file
+    // to read in as the identifier for whether or not we have already read the
+    // requested data. Doing this at runtime avoids having to collect all
+    // data files that will be read in at compile-time to initialize a flag in
+    // the DataBox for each of them.
+    const auto& has_read_volume_data = db::get<Tags::HasReadVolumeData>(box);
+    const auto volume_data_id = pretty_type::get_name<ImporterOptionsGroup>();
```

Comment:
I'd like to double-check I understand what this is doing. Can you please give me an example of what ImporterOptionsGroup's type is, the corresponding value of volume_data_id, and how the next line checks to make sure the data has been read? I would have thought ImporterOptionsGroup contained all of the options from the group.

I guess my real question is why not just set a bool here?

Reply:
The importers dev guide (in this PR) has an example at the bottom. The `ImporterOptionsGroup` is the location in the input file from where we read the filename, subfile and observation ID, so it identifies the file to read in. The `volume_data_id` is just a string representation of the class name, so again, it identifies the file to read in. This ID is placed in the DataBox to say "we have read this file", so subsequent calls don't read it again. To make this a bool we would have to collect all possible option groups (i.e. files) that will be read in, so we can initialize a bool in the DataBox for each of the files.